File: crm/app/internal/adapters/inbound/web/screen_inbox.go
// Package web — S05 Inbox + S06 Conversation Detail screen handlers.
// Inbound adapter: reads/writes via ConversationLister, ConvWriter, ConvLinker,
// AppUserLister, PartyLister, ActivityLogger. No business logic here.
package web

import (
	"log"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"

	"github.com/vantt/data-integration/crm/app/internal/adapters/inbound/web/templates"
	"github.com/vantt/data-integration/crm/app/internal/domain"
)

func init() {
	register(func(r chi.Router, d *Deps) {
		// S05 — Inbox list.
		r.Get("/inbox", d.handleInbox)

		// S06 — Conversation detail.
		r.Get("/inbox/{conversationID}", d.handleConversationDetail)

		// M09 — Assign conversation modal + submit.
		r.Get("/inbox/{conversationID}/modal/assign", d.handleModalAssignConversation)
		r.Post("/inbox/{conversationID}/assign", d.handleAssignConversation)

		// M10 — Close conversation modal + submit.
		r.Get("/inbox/{conversationID}/modal/close", d.handleModalCloseConversation)
		r.Post("/inbox/{conversationID}/close", d.handleCloseConversation)

		// M11 — Link party modal + party search fragment + submit.
		r.Get("/inbox/{conversationID}/modal/link-party", d.handleModalLinkParty)
		r.Get("/inbox/{conversationID}/parties/search", d.handleInboxPartySearch)
		r.Post("/inbox/{conversationID}/link-party", d.handleLinkParty)

		// M08 — Log activity from S06.
		r.Get("/inbox/{conversationID}/modal/log-activity", d.handleModalLogActivityConv)
		r.Post("/inbox/{conversationID}/log-activity", d.handleLogActivityOnConv)
	})
}

// ── S05 Inbox ─────────────────────────────────────────────────────────────────

func (d *Deps) handleInbox(w http.ResponseWriter, r *http.Request) {
	status := r.URL.Query().Get("status")
	assignee := r.URL.Query().Get("assignee")

	convs, err := d.Conversations.ListInbox(r.Context(), assignee, status)
	if err != nil {
		log.Printf("inbox: list: %v", err)
		convs = []domain.Conversation{}
	}

	// Build partyID → displayName for rows that have a linked party.
	partyNames := d.convPartyNames(r, convs)

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	if err := templates.InboxPage(convs, partyNames, status, assignee).Render(r.Context(), w); err != nil {
		log.Printf("inbox: render: %v", err)
	}
}

// ── S06 Conversation Detail ────────────────────────────────────────────────────

func (d *Deps) handleConversationDetail(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "conversationID")

	conv, err := d.ConvLinker.GetByID(r.Context(), convID)
	if err != nil || conv == nil {
		http.Error(w, "Không tìm thấy hội thoại", http.StatusNotFound)
		return
	}

	msgs, err := d.Conversations.GetMessages(r.Context(), convID)
	if err != nil {
		log.Printf("conv detail: get messages %s: %v", convID, err)
		msgs = []domain.Message{}
	}

	// Sidebar: resolve linked party name + 360 link.
	var partyName, partyLink string
	if conv.PartyID != nil {
		if p, err2 := d.Parties.GetByID(r.Context(), *conv.PartyID); err2 == nil && p != nil {
			partyName = p.DisplayName
			partyLink = "/customers/" + p.PartyID
		}
	}

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	if err := templates.ConversationDetailPage(conv, msgs, partyName, partyLink).Render(r.Context(), w); err != nil {
		log.Printf("conv detail: render %s: %v", convID, err)
	}
}

// ── M09 Assign Conversation ───────────────────────────────────────────────────

func (d *Deps) handleModalAssignConversation(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "conversationID")
	conv, err := d.ConvLinker.GetByID(r.Context(), convID)
	if err != nil || conv == nil {
		http.Error(w, "not found", http.StatusNotFound)
		return
	}
	users, _ := d.AppUsers.ListActive(r.Context())

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	if err := templates.ModalAssignConversation(convID, conv, users).Render(r.Context(), w); err != nil {
		log.Printf("modal assign: render: %v", err)
	}
}

func (d *Deps) handleAssignConversation(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "conversationID")
	if err := r.ParseForm(); err != nil {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}
	assigneeID := strings.TrimSpace(r.FormValue("assignee_user_id"))
	if assigneeID == "" {
		http.Error(w, "assignee_user_id required", http.StatusBadRequest)
		return
	}
	if err := d.ConvWriter.AssignConversation(r.Context(), convID, assigneeID); err != nil {
		log.Printf("assign conv %s: %v", convID, err)
		http.Error(w, "failed to assign", http.StatusInternalServerError)
		return
	}
	w.Header().Set("HX-Trigger", `{"closeModal":true}`)
	w.WriteHeader(http.StatusOK)
}

// ── M10 Close Conversation ────────────────────────────────────────────────────

func (d *Deps) handleModalCloseConversation(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "conversationID")
	conv, err := d.ConvLinker.GetByID(r.Context(), convID)
	if err != nil || conv == nil {
		http.Error(w, "not found", http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	if err := templates.ModalCloseConversation(convID, conv).Render(r.Context(), w); err != nil {
		log.Printf("modal close conv: render: %v", err)
	}
}

func (d *Deps) handleCloseConversation(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "conversationID")
	if err := r.ParseForm(); err != nil {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}
	if err := d.ConvWriter.SetStatus(r.Context(), convID, "closed"); err != nil {
		log.Printf("close conv %s: %v", convID, err)
		http.Error(w, "failed to close", http.StatusInternalServerError)
		return
	}
	w.Header().Set("HX-Trigger", `{"closeModal":true}`)
	w.Header().Set("HX-Redirect", "/inbox/"+convID)
	w.WriteHeader(http.StatusOK)
}

// ── M11 Link Party to Conversation ───────────────────────────────────────────

func (d *Deps) handleModalLinkParty(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "conversationID")
	conv, err := d.ConvLinker.GetByID(r.Context(), convID)
	if err != nil || conv == nil {
		http.Error(w, "not found", http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	if err := templates.ModalLinkParty(convID, conv).Render(r.Context(), w); err != nil {
		log.Printf("modal link party: render: %v", err)
	}
}

// handleInboxPartySearch serves GET /inbox/{conversationID}/parties/search?q=… — live search fragment inside M11.
func (d *Deps) handleInboxPartySearch(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "conversationID")
	q := strings.TrimSpace(r.URL.Query().Get("q"))
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	if q == "" {
		_, _ = w.Write([]byte(`<div class="text-muted">Nhập tên hoặc SĐT để tìm...</div>`))
		return
	}
	partyIDs, err := d.Parties.SearchByName(r.Context(), q)
	if err != nil {
		log.Printf("inbox party search: %v", err)
		partyIDs = nil
	}
	parties := d.fetchPartiesByIDs(r, partyIDs)
	// Pass convID so each result row's hx-post targets the correct /inbox/{convID}/link-party URL.
	if err := templates.PartySearchResults(parties, convID).Render(r.Context(), w); err != nil {
		log.Printf("inbox party search: render: %v", err)
	}
}

func (d *Deps) handleLinkParty(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "conversationID")
	if err := r.ParseForm(); err != nil {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}
	partyID := strings.TrimSpace(r.FormValue("party_id"))
	if partyID == "" {
		http.Error(w, "party_id required", http.StatusBadRequest)
		return
	}
	if err := d.ConvLinker.LinkParty(r.Context(), convID, partyID); err != nil {
		log.Printf("link party conv %s party %s: %v", convID, partyID, err)
		http.Error(w, "failed to link party", http.StatusInternalServerError)
		return
	}
	// Full redirect so the sidebar re-renders with party info.
	w.Header().Set("HX-Redirect", "/inbox/"+convID)
	w.WriteHeader(http.StatusOK)
}

// ── M08 Log Activity (from S06) ───────────────────────────────────────────────

func (d *Deps) handleModalLogActivityConv(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "conversationID")
	conv, err := d.ConvLinker.GetByID(r.Context(), convID)
	if err != nil || conv == nil {
		http.Error(w, "not found", http.StatusNotFound)
		return
	}
	var partyID string
	if conv.PartyID != nil {
		partyID = *conv.PartyID
	}
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	if err := templates.ModalLogActivity(partyID, convID).Render(r.Context(), w); err != nil {
		log.Printf("modal log activity conv: render: %v", err)
	}
}

func (d *Deps) handleLogActivityOnConv(w http.ResponseWriter, r *http.Request) {
	convID := chi.URLParam(r, "conversationID")
	if err := r.ParseForm(); err != nil {
		http.Error(w, "bad request", http.StatusBadRequest)
		return
	}
	partyID := strings.TrimSpace(r.FormValue("party_id"))
	actType := strings.TrimSpace(r.FormValue("activity_type"))
	body := strings.TrimSpace(r.FormValue("body"))
	if partyID == "" || body == "" {
		http.Error(w, "party_id and body required", http.StatusBadRequest)
		return
	}
	if actType == "" {
		actType = "chat"
	}
	d.logActivityShared(w, r, partyID, actType, body, convID)
}

// ── shared helpers ────────────────────────────────────────────────────────────

// convPartyNames builds partyID → displayName for all conversations that have a party.
func (d *Deps) convPartyNames(r *http.Request, convs []domain.Conversation) map[string]string {
	out := make(map[string]string)
	for _, c := range convs {
		if c.PartyID == nil {
			continue
		}
		if _, seen := out[*c.PartyID]; seen {
			continue
		}
		if p, err := d.Parties.GetByID(r.Context(), *c.PartyID); err == nil && p != nil {
			out[*c.PartyID] = p.DisplayName
		}
	}
	return out
}

// fetchPartiesByIDs resolves a slice of party IDs to domain.Party values.
func (d *Deps) fetchPartiesByIDs(r *http.Request, ids []string) []domain.Party {
	out := make([]domain.Party, 0, len(ids))
	for _, id := range ids {
		if p, err := d.Parties.GetByID(r.Context(), id); err == nil && p != nil {
			out = append(out, *p)
		}
	}
	return out
}

// logActivityShared logs an activity via ActivityLogger; used from S06 and S03.
func (d *Deps) logActivityShared(w http.ResponseWriter, r *http.Request, partyID, actType, body, convID string) {
	a := &domain.Activity{
		PartyID:      partyID,
		ActivityType: actType,
		Body:         body,
		Channel:      "messenger",
	}
	if err := d.ActivityLog.LogActivity(r.Context(), a); err != nil {
		log.Printf("log activity party %s: %v", partyID, err)
		http.Error(w, "failed to log activity", http.StatusInternalServerError)
		return
	}
	w.Header().Set("HX-Trigger", `{"closeModal":true}`)
	if convID != "" {
		w.Header().Set("HX-Redirect", "/inbox/"+convID)
	}
	w.WriteHeader(http.StatusOK)
}

File: crm/src/adapters/inbound/web/screen_inbox.py
"""Web adapter — S05 Inbox + S06 Conversation Detail screens.

FastAPI router mirroring screen_inbox.go.
Serves full inbox list page and conversation detail page/fragment.
No business logic — thin adapter only.
"""
from __future__ import annotations

import logging
from typing import Optional, Protocol

from fastapi import APIRouter, Form, Request, Response
from fastapi.responses import HTMLResponse
from fastapi.templating import Jinja2Templates

from crm.src.domain.entities.conversation import Conversation, Message
from crm.src.domain.entities.app_user import AppUser
from crm.src.domain.entities.party import Party

log = logging.getLogger(__name__)

# ── Service protocols ─────────────────────────────────────────────────────────

class ConversationLister(Protocol):
    def list_inbox(self, assignee: str, status: str) -> list[Conversation]: ...
    def get_messages(self, conv_id: str) -> list[Message]: ...


class ConvReader(Protocol):
    def get_by_id(self, conv_id: str) -> Optional[Conversation]: ...


class ConvWriter(Protocol):
    def assign_conversation(self, conv_id: str, assignee_user_id: str) -> None: ...
    def set_status(self, conv_id: str, status: str) -> None: ...
    def link_party(self, conv_id: str, party_id: str) -> None: ...


class PartyReader(Protocol):
    def get_by_id(self, party_id: str) -> Optional[Party]: ...
    def search_by_name(self, q: str) -> list[str]: ...


class AppUserLister(Protocol):
    def list_active(self) -> list[AppUser]: ...


class ActivityLogger(Protocol):
    def log_activity(self, activity_data: dict) -> object: ...


# ── Helpers ───────────────────────────────────────────────────────────────────

def _conv_status_bdg(status: str) -> str:
    return {"open": "bdg--good", "pending": "bdg--warn", "closed": ""}.get(status, "")


def _conv_party_names(
    convs: list[Conversation], party_reader: PartyReader
) -> dict[str, str]:
    """Build partyID → displayName for all conversations that have a linked party."""
    out: dict[str, str] = {}
    for c in convs:
        if c.party_id is None or c.party_id in out:
            continue
        try:
            p = party_reader.get_by_id(c.party_id)
            if p:
                out[c.party_id] = p.display_name
        except Exception:
            pass
    return out


# ── Router factory ────────────────────────────────────────────────────────────

def make_inbox_router(
    templates: Jinja2Templates,
    conversations: ConversationLister,
    conv_reader: ConvReader,
    conv_writer: ConvWriter,
    parties: PartyReader,
    app_users: AppUserLister,
    activity_log: ActivityLogger,
) -> APIRouter:
    """Return APIRouter wired with all Inbox + Conversation Detail routes."""
    router = APIRouter()

    # ── S05 Inbox list ────────────────────────────────────────────────────────

    @router.get("/inbox", response_class=HTMLResponse)
    async def handle_inbox(request: Request) -> Response:
        status = request.query_params.get("status", "")
        assignee = request.query_params.get("assignee", "")
        try:
            convs = conversations.list_inbox(assignee, status)
        except Exception as exc:
            log.error("inbox: list: %s", exc)
            convs = []
        party_names = _conv_party_names(convs, parties)
        return templates.TemplateResponse(
            "inbox.html",
            {
                "request": request,
                "convs": convs,
                "party_names": party_names,
                "status_filter": status,
                "assignee_filter": assignee,
                "conv_status_bdg": _conv_status_bdg,
            },
        )

    # ── S06 Conversation detail ───────────────────────────────────────────────

    @router.get("/inbox/{conv_id}", response_class=HTMLResponse)
    async def handle_conversation_detail(request: Request, conv_id: str) -> Response:
        conv = conv_reader.get_by_id(conv_id)
        if conv is None:
            return HTMLResponse("Không tìm thấy hội thoại", status_code=404)
        try:
            msgs = conversations.get_messages(conv_id)
        except Exception as exc:
            log.error("conv detail: get messages %s: %s", conv_id, exc)
            msgs = []
        party_name, party_link = "", ""
        if conv.party_id:
            try:
                p = parties.get_by_id(conv.party_id)
                if p:
                    party_name = p.display_name
                    party_link = f"/customers/{p.party_id}"
            except Exception:
                pass
        return templates.TemplateResponse(
            "conversation_detail.html",
            {
                "request": request,
                "conv": conv,
                "msgs": msgs,
                "party_name": party_name,
                "party_link": party_link,
                "conv_status_bdg": _conv_status_bdg,
            },
        )

    # ── M09 Assign conversation ───────────────────────────────────────────────

    @router.get("/inbox/{conv_id}/modal/assign", response_class=HTMLResponse)
    async def handle_modal_assign(request: Request, conv_id: str) -> Response:
        conv = conv_reader.get_by_id(conv_id)
        if conv is None:
            return HTMLResponse("not found", status_code=404)
        try:
            users = app_users.list_active()
        except Exception:
            users = []
        return templates.TemplateResponse(
            "fragments/modal_assign_conversation.html",
            {"request": request, "conv_id": conv_id, "conv": conv, "users": users},
        )

    @router.post("/inbox/{conv_id}/assign", response_class=HTMLResponse)
    async def handle_assign_conversation(
        request: Request, conv_id: str, assignee_user_id: str = Form(...)
    ) -> Response:
        assignee_user_id = assignee_user_id.strip()
        if not assignee_user_id:
            return HTMLResponse("assignee_user_id required", status_code=400)
        try:
            conv_writer.assign_conversation(conv_id, assignee_user_id)
        except Exception as exc:
            log.error("assign conv %s: %s", conv_id, exc)
            return HTMLResponse("failed to assign", status_code=500)
        return Response(status_code=200, headers={"HX-Trigger": '{"closeModal":true}'})

    # ── M10 Close conversation ────────────────────────────────────────────────

    @router.get("/inbox/{conv_id}/modal/close", response_class=HTMLResponse)
    async def handle_modal_close(request: Request, conv_id: str) -> Response:
        conv = conv_reader.get_by_id(conv_id)
        if conv is None:
            return HTMLResponse("not found", status_code=404)
        return templates.TemplateResponse(
            "fragments/modal_close_conversation.html",
            {"request": request, "conv_id": conv_id, "conv": conv},
        )

    @router.post("/inbox/{conv_id}/close", response_class=HTMLResponse)
    async def handle_close_conversation(request: Request, conv_id: str) -> Response:
        try:
            conv_writer.set_status(conv_id, "closed")
        except Exception as exc:
            log.error("close conv %s: %s", conv_id, exc)
            return HTMLResponse("failed to close", status_code=500)
        return Response(
            status_code=200,
            headers={"HX-Trigger": '{"closeModal":true}', "HX-Redirect": f"/inbox/{conv_id}"},
        )

    # ── M11 Link party ────────────────────────────────────────────────────────

    @router.get("/inbox/{conv_id}/modal/link-party", response_class=HTMLResponse)
    async def handle_modal_link_party(request: Request, conv_id: str) -> Response:
        conv = conv_reader.get_by_id(conv_id)
        if conv is None:
            return HTMLResponse("not found", status_code=404)
        return templates.TemplateResponse(
            "fragments/modal_link_party.html",
            {"request": request, "conv_id": conv_id, "conv": conv},
        )

    @router.get("/inbox/{conv_id}/parties/search", response_class=HTMLResponse)
    async def handle_inbox_party_search(request: Request, conv_id: str) -> Response:
        q = request.query_params.get("q", "").strip()
        if not q:
            return HTMLResponse('<div class="text-muted">Nhập tên hoặc SĐT để tìm...</div>')
        try:
            ids = parties.search_by_name(q)
            found = [p for pid in ids for p in [parties.get_by_id(pid)] if p]
        except Exception as exc:
            log.error("inbox party search: %s", exc)
            found = []
        return templates.TemplateResponse(
            "fragments/party_search_results.html",
            {"request": request, "parties": found, "conv_id": conv_id},
        )

    @router.post("/inbox/{conv_id}/link-party", response_class=HTMLResponse)
    async def handle_link_party(
        request: Request, conv_id: str, party_id: str = Form(...)
    ) -> Response:
        party_id = party_id.strip()
        if not party_id:
            return HTMLResponse("party_id required", status_code=400)
        try:
            conv_writer.link_party(conv_id, party_id)
        except Exception as exc:
            log.error("link party conv %s party %s: %s", conv_id, party_id, exc)
            return HTMLResponse("failed to link party", status_code=500)
        return Response(status_code=200, headers={"HX-Redirect": f"/inbox/{conv_id}"})

    # ── M08 Log activity (from S06) ───────────────────────────────────────────

    @router.get("/inbox/{conv_id}/modal/log-activity", response_class=HTMLResponse)
    async def handle_modal_log_activity(request: Request, conv_id: str) -> Response:
        conv = conv_reader.get_by_id(conv_id)
        if conv is None:
            return HTMLResponse("not found", status_code=404)
        party_id = conv.party_id or ""
        return templates.TemplateResponse(
            "fragments/modal_log_activity.html",
            {"request": request, "party_id": party_id, "conv_id": conv_id},
        )

    @router.post("/inbox/{conv_id}/log-activity", response_class=HTMLResponse)
    async def handle_log_activity_on_conv(
        request: Request,
        conv_id: str,
        party_id: str = Form(...),
        activity_type: str = Form(default="chat"),
        body: str = Form(...),
    ) -> Response:
        party_id = party_id.strip()
        body = body.strip()
        if not party_id or not body:
            return HTMLResponse("party_id and body required", status_code=400)
        current_user = getattr(request.state, "current_user", None)
        try:
            activity_log.log_activity({
                "party_id": party_id,
                "activity_type": activity_type or "chat",
                "body": body,
                "channel": "messenger",
                "direction": "out",
                "staff_user_id": current_user.user_id if current_user else None,
            })
        except Exception as exc:
            log.error("log activity party %s: %s", party_id, exc)
            return HTMLResponse("failed to log activity", status_code=500)
        return Response(
            status_code=200,
            headers={"HX-Trigger": '{"closeModal":true}', "HX-Redirect": f"/inbox/{conv_id}"},
        )

    return router
